### PR TITLE
expression: fix incorrect collation when cast non-string type arg to string type

### LINKED ISSUE #19189: result of select collation(concat(1,1)); is wrong
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
set names utf8mb4 collate utf8mb4_general_ci;
select collation(concat(1,1));
select collation(concat(1 collate `binary`));
```

### 2. What did you expect to see? (Required)
`utf8mb4_general_ci`
`binary`

### 3. What did you see instead (Required)
`utf8mb4_bin`
`utf8mb4_bin`
### 
this bug was introduced after https://github.com/pingcap/tidb/pull/19169. but it is not the root cause, since before this PR 
```sql
select collation(concat(1 collate `binary`));
```
is wrong too

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close #19189 #19166 
Problem Summary:

### What is changed and how it works?

- when non-string type args cast to string type, it should set collation to the origin collation.
- `deriveCoercibilityForScarlarFunc` should return `CoercibilityIgnorable` if all args is `null`

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- fix an issue that may get wrong collation result

### CODE REVIEW

File: expression/collation.go [hunk 1]
```diff
@@ -143,12 +143,17 @@ func deriveCoercibilityForScarlarFunc(sf *ScalarFunction) Coercibility {
...
 	}
-	coer := CoercibilityCoercible
+	coer := CoercibilityIgnorable
 	for _, arg := range sf.GetArgs() {
 		if arg.Coercibility() < coer {
 			coer = arg.Coercibility()
 		}
 	}
+
+	if coer == CoercibilityNumeric {
```
Comment (wjhuang2016) [inline expression/collation.go:153]: Why set CoercibilityCoercible? Please add a comment for it.

**Comment by xiongjiwei** [comment]
@qw4990 could please help me merge this PR?

**Pull request merged**